src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala

```diff
@@ -0,0 +1,423 @@
+package mesosphere.mesos
+
+import mesosphere.marathon.MarathonConf
+import mesosphere.marathon.core.instance.Instance
+import mesosphere.marathon.core.pod.PodDefinition
+import mesosphere.marathon.core.task.Task
+import mesosphere.marathon.raml
+import mesosphere.marathon.state.{ EnvVarString, PathId, Timestamp }
+import mesosphere.marathon.tasks.PortsMatch
+import mesosphere.mesos.ResourceMatcher.ResourceSelector
+import org.apache.mesos.{ Protos => mesos }
+import org.slf4j.LoggerFactory
+
+import scala.collection.JavaConverters._
+import scala.collection.immutable.Seq
+
+// TODO(PODS): Integrate plugin support (e.g. secrets)
+object TaskGroupBuilder {
+  val log = LoggerFactory.getLogger(getClass)
+
+  def build(
+    podDefinition: PodDefinition,
+    offer: mesos.Offer,
+    runningTasks: => Iterable[Task],
+    newInstanceId: PathId => Instance.Id,
+    config: MarathonConf
+  ): Option[(mesos.ExecutorInfo, mesos.TaskGroupInfo, Seq[Option[Int]])] = {
+    val acceptedResourceRoles: Set[String] = {
+      val roles = if (podDefinition.acceptedResourceRoles.isEmpty) {
+        config.defaultAcceptedResourceRolesSet
+      } else {
+        podDefinition.acceptedResourceRoles
+      }
+      if (log.isDebugEnabled) log.debug(s"acceptedResourceRoles $roles")
+      roles
+    }
+
+    val resourceMatchOpt: Option[ResourceMatcher.ResourceMatch] =
+      ResourceMatcher.matchResources(offer, podDefinition, runningTasks, ResourceSelector.any(acceptedResourceRoles))
+
+    resourceMatchOpt match {
```

Comment:
But it then also needs a `None` case so it's `resourceMatchOpt.map(build(podDefinition, offer, newInstanceId, config, _)).getOrElse(None)`